### PR TITLE
[SimplifyLibCalls] Constant fold nan libcall

### PR DESCRIPTION
**Pull request by dtcxzyw**
Reference: https://en.cppreference.com/w/c/numeric/math/nan
The logic is copied from clang frontend:

https://github.com/llvm/llvm-project/blob/1d2b2d29d733200b704f38d220d22ecc07d6cf42/clang/lib/AST/ExprConstant.cpp#L14741-L14777

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/nan.ll [hunk 1]
```diff
@@ -0,0 +1,41 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+@empty = private unnamed_addr constant [1 x i8] zeroinitializer, align 1
+@dec = private unnamed_addr constant [2 x i8] c"1\00", align 1
+@hex = private unnamed_addr constant [4 x i8] c"0xf\00", align 1
+
+define double @nan_empty() {
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/nan.ll:8]: Check the undef/poison/null, non-constant cases?

**Review by arsenm — COMMENTED** [review]
Unrelated, but in the attribute inference you could mark these returns as nofpclass(~nan)

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 2]
```diff
@@ -3858,6 +3858,22 @@ Value *LibCallSimplifier::optimizeStringMemoryLibCall(CallInst *CI,
   return nullptr;
 }
 
+/// Constant folding nan/nanf/nanl.
+static Value *optimizeNaN(CallInst *CI) {
+  StringRef CharSeq;
+  if (!getConstantStringInfo(CI->getArgOperand(0), CharSeq))
+    return nullptr;
+
+  llvm::APInt Fill;
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:3867]:
```suggestion
  APInt Fill;
```

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp [hunk 3]
```diff
@@ -3858,6 +3858,22 @@ Value *LibCallSimplifier::optimizeStringMemoryLibCall(CallInst *CI,
...
+  // Treat empty strings as if they were zero.
+  if (CharSeq.empty())
+    Fill = llvm::APInt(32, 0);
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:3870]:
```suggestion
    Fill = APInt(32, 0);
```

File: llvm/test/Transforms/InstCombine/nan.ll [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+define double @nan_null() {
+; CHECK-LABEL: define double @nan_null() {
+; CHECK-NEXT:    [[RES:%.*]] = call double @nan(ptr null)
+; CHECK-NEXT:    ret double [[RES]]
+;
+  %res = call double @nan(ptr null)
+  ret double %res
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/nan.ll:68]: Should have a non-constant negative test too

**Review by arsenm — APPROVED** [review]

**Pull request merged**